src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] array;
+    private int size;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            T[] newArray = (T[]) new Object[(int) (array.length * 1.5)];
+            System.arraycopy(array, 0, newArray, 0, array.length);
+            array = newArray;
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if ((index < 0) || (index > size)) {
+            throw new ArrayListIndexOutOfBoundsException("The index extends beyond the array.");
```

Comment:
The index check is repeated in multiple methods (`add`, `get`, `set`, `remove`). To avoid code repetition, consider creating a private method that ensures the index is within the valid range.